metagraph/src/annotation/row_diff_builder.cpp

```diff
@@ -583,76 +583,76 @@ void optimize_anchors_in_row_diff(const std::string &graph_fname,
+                logger->error("Row reduction vectors have different sizes");
+                exit(1);
+            }
+            if (row_reduction.back().width() != row_reduction.front().width()) {
+                logger->error("Row reduction vectors have different width: {} != {}",
+                              row_reduction.back().width(),
+                              row_reduction.front().width());
+                exit(1);
+            }
         }
     }
 
-    if (!filenames.size()) {
+    if (!row_reduction.size()) {
         logger->error("Didn't find any row reduction vectors in {} to merge", dest_dir);
         exit(1);
     }
 
-    if (filenames.size() > 1u)
-        logger->trace("Merging row reduction vectors");
-
-    while (filenames.size() > 1u) {
-        std::vector<std::string> filenames_new((filenames.size() + 1) / 2);
+    sdsl::bit_vector anchors;
+    uint64_t num_anchors_old;
+    {
+        anchor_bv_type old_anchors;
+        auto original_anchors_fname = graph_fname + kRowDiffAnchorExt + ".unopt";
 
-        #pragma omp parallel for num_threads(get_num_threads()) schedule(dynamic)
-        for (size_t t = 1; t < filenames.size(); t += 2) {
-            // compute sum of t-1 and t.
-            sdsl::int_vector_buffer first(filenames[t - 1], std::ios::in, 1024 * 1024);
-            sdsl::int_vector_buffer second(filenames[t], std::ios::in, 1024 * 1024);
-            if (first.size() != second.size() || first.width() != second.width()) {
-                logger->error("Sizes of row reduction vectors are incompatible, {}: {}, {}: {}",
-                              filenames[t - 1], first.size(), filenames[t], second.size());
-                exit(1);
-            }
-            filenames_new[t / 2] = fmt::format("{}.merged", filenames[t]);
-            sdsl::int_vector_buffer sum(filenames_new[t / 2], std::ios::out, 1024 * 1024, first.width());
+        std::ifstream f(original_anchors_fname, ios::binary);
+        old_anchors.load(f);
 
-            for (uint64_t i = 0; i < first.size(); ++i) {
-                sum.push_back(first[i] + second[i]);
-            }
+        if (old_anchors.size() != row_reduction.front().size()) {
+            logger->error(
+                "Original anchors {}: {} and row reduction vectors: {} are incompatible",
+                original_anchors_fname, old_anchors.size(), row_reduction.front().size());
+            exit(1);
         }
+        num_anchors_old = old_anchors.num_set_bits();
+        anchors = old_anchors.convert_to<sdsl::bit_vector>();
+    }
 
-        if (filenames.size() % 2)
-            filenames_new.back() = filenames.back();
-
-        logger->trace("Merged {} row reduction vectors into {}",
-                      filenames.size(), filenames_new.size());
+    std::vector<uint64_t> buf;
+    static_assert(sizeof(uint64_t) * 8 > ROW_REDUCTION_WIDTH);
+    const uint64_t MINUS_BIT = 1llu << (row_reduction.front().width() - 1);
```

Comment:
This won't always work, since you can still have an overflow and have this bit be 0. For example, if width = 2 and the value is 8, the check below will fail.

```suggestion
    const uint64_t OVERFLOW_MASK = 0xFFFFFFFFFFFFFFFF << row_reduction.front().width();
```

Reply:
Unless I misunderstood something and this is checking something else.

Reply:
I think it's good, yours won't work for negative arithmetics. Could you double check please?

Reply:
alternately, one could sign-extend the values read from `row_reduction` to get `uint64_t`, then do the addition, and finally check the most significant bit to see if it's set

Reply:
so the summation line would look something like
```c++
static_cast<uint64_t>(static_cast<int64_t>(static_cast<int32_t>(row_reduction[j][i + t])))
```
but this looks ugly...

Reply:
Is there any problem with the current code?

Reply:
Ah, you suggest to cast it to signed/unsigned, and then one doesn't have to bother about the internals.

A downside is that it would work only with 32 bits, while the current thing is meant to work with arbitrary width.

Though maybe it's fine to hardcode this... Should we do that?

Reply:
so a minimal example of the problem I'm talking about looks like this
```c++
#include <cstdint>
#include <limits>
#include <iostream>

int main() {
    uint8_t width = 32;
    uint64_t a = 1llu << (width - 1);
    uint64_t b = a + a;

    uint64_t mask = 1llu << (width - 1);
    std::cout << (b & mask) << "\n";
    return 0;
}
```
which outputs 0

Reply:
but fixing this for arbitrary widths becomes a bit complicated because we can't use casts to do the conversion

Reply:
What is the problem with the current code? Isn't this value bounded by the no of columns? We are many orders of magnitude below overflowing an int32, there is no way we can overflow an int64.

Reply:
It won't happen given the current sizes, but if we ever change the width to something smaller, we might run into this issue.

Reply:
Harun, your example shows the overflow issue, not the incorrectness of arithmetics.

The overflow is unavoidable since we use a finite number of bits for numbres.

The thing we must make correct though is the emulation of negative numbers and their addition.

Like, `A + B - A = B` for all |A|, |B| < 2^31.

This means, for instance, that the mask has to be `1llu << (row_reduction.front().width() - 1);` and not `const uint64_t OVERFLOW_MASK = 0xFFFFFFFFFFFFFFFF << row_reduction.front().width();`.
Because of the following:
```c++
delta[0] = 0;  // 0x00000000
delta[0] -= 1;  // 0xFFFFFFFF
uint64_t x = delta[0]; // 0x00000000FFFFFFFF  (because of truncation with 32 bits)
x += 1; // 0x0000000100000000  (because of truncation with 32 bits)
x & OVERFLOW_MASK != 0 // so it looks like x<0, which is wrong result, because the actual result should be 0.
```